README

 MALHEUR - Automatic Analysis of Malware Behavior
 Copyright (c) 2009-2012 Konrad Rieck (konrad@mlsec.org)          
 University of Goettingen, Berlin Institute of Technology

 Introduction 
 --

 Malheur is a tool for the automatic analysis of malware behavior (program
 behavior recorded from malicious software in a sandbox environment).  It
 has been designed to support the regular analysis of malicious software and
 the development of detection and defense measures.  Malheur allows for
 identifying novel classes of malware with similar behavior and assigning
 unknown malware to discovered classes.  It supports four basic actions for
 analysis which can be applied to reports of recorded behavior:

 1. Extraction of prototypes
     From a given set of reports, malheur identifies a subset of
     prototypes representative for the full data set. The prototypes
     provide a quick overview of recorded behavior and can be used to
     guide manual inspection.

 2. Clustering of behavior
     Malheur automatically identifies groups (clusters) of reports
     containing similar behavior. Clustering allows for discovering novel
     classes of malware and provides the basis for crafting specific
     detection and defense mechanisms, such as anti-virus signatures.

 3. Classification of behavior
     Based on a set of previously clustered reports, malheur is able to
     assign unknown behavior to known groups of malware. Classification
     enables identifying novel and unknown variants of malware and can be
     used to filter program behavior prior to manual inspection.

 4. Incremental analysis
     Malheur can be applied incrementally for analysis of large data
     sets. By processing reports in chunks, the run-time as well as
     memory requirements can be significantly reduced. This renders
     long-term application of malheur feasible, for example for daily
     analysis of incoming malware programs.

 A detailed description of these techniques as well as technical
 background on analysis of malicious software is provided in the
 following articles:

     "Automatic Analysis of Malware Behavior using Machine Learning"
     Konrad Rieck, Philipp Trinius, Carsten Willems, and Thorsten Holz
     Journal of Computer Security (JCS), 19 (4) 639-668, 2011.

     "A Malware Instruction Set for Behavior-Based Analysis" 
     Philipp Trinius, Carsten Willems, Thorsten Holz, and Konrad Rieck 
     Technical report TR-2009-07, University of Mannheim, 2009

 Dependencies
 --
      >= libconfig-1.4
      >= libarchive-2.70

 Compilation & Installation
 --
 
 From GIT repository first run

      $ ./bootstrap

 From tarball run

      $ ./configure [options]
      $ make
      $ make check
      $ make install

 Options for configure

      --prefix=PATH           Set directory prefix for installation

      By default Malheur is installed into /usr/local. If you prefer
      a different location, use this option to select an installation
      directory.

      --enable-openmp         Enable support for OpenMP (experimental)
    
      This option enables support for OpenMP in Malheur. Several
      functions of the malware analysis have been enhanced using
      OpenMP directives, such that they execute in parallel and
      benefit from multi-core architectures.

      --enable-matlab         Enable optional Matlab tools
      --with-matlab-dir=PATH  Set directory prefix of matlab installation

      Some functions of Malheur are also available in form of Matlab
      .mex files which allows for using implemented analysis
      methods directly from within a Matlab environment.

 License
 --

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 3 of the License, or
 (at your option) any later version.  This program is distributed
 without any warranty. See the GNU General Public License for more
 details.


CHANGES
2012-12-27  Konrad Rieck <konrad@mlsec.org>

    - fixed annoying ptr warning

    - updated autotool config

    - changed order of headers

2012-05-26  Konrad Rieck <konrad@mlsec.org>

    - added md README

2012-02-05  Konrad Rieck <konrad@mlsec.org>

    - improved loading of state. started support for info

    - changed working dir of Malheur to local-state dir

    - Malheur development continues in Goettingen now

2012-01-22  Konrad Rieck <konrad@mlsec.org>

    - minor changes to env scripts

2011-09-23  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - sync

2011-08-25  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - rewrote default configuration code

2011-08-24  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - fixed another libconfig related bug.

    - fixed problem with FLT_NONE constant. it is not simply NAN

2011-07-09  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - sync

    - missing files

2011-07-04  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - fixed printing of configuration

2011-04-19  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - updated manual page

2011-04-12  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - missed one release, so back to previous version number

2011-03-29  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - added support for longopts

2010-12-09  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - some new todos

2010-10-08  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - updated documentaiton

    - support for computing prototype distances

2010-10-07  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - updated documentation and added changes file

    - improved output of shared n-grams

2010-09-16  Konrad Rieck <rieck@node16.ml.tu-berlin.de>

    - switched to libconfig 1.4.x

2010-08-06  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - updated man page

    - changed email address

2010-08-04  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - improved computation of shared n-grams

2010-08-03  Konrad Rieck <konrad.rieck@tu-berlin.de>

    - support for shared ngrams

2010-02-16  Konrad Rieck <rieck@cs.tu-berlin.de>

    - fixed bug with non-deterministic selection of first prototype

2010-02-03  Konrad Rieck <rieck@cs.tu-berlin.de>

    - added support to disable saving

2010-01-16  Konrad Rieck <rieck@cs.tu-berlin.de>

    - general code clean-up (casting, printf formats, etc)

2010-01-13  Konrad Rieck <rieck@cs.tu-berlin.de>

    - added local uthash version as fallback

2010-01-11  Konrad Rieck <rieck@cs.tu-berlin.de>

    - removed internal uthash

2010-01-01  Konrad Rieck <rieck@cs.tu-berlin.de>

    - fixed bug in classification mode

2009-12-28  Konrad Rieck <rieck@cs.tu-berlin.de>

    - improved documentation and examples

2009-12-16  Konrad Rieck <rieck@cs.tu-berlin.de>

    - fixed installation problems

2009-12-12  Konrad Rieck <rieck@cs.tu-berlin.de>

    - minor fixes to math. normalized vectors are mandatory

    - changed MIST format to latest

    - minor fixes

2009-12-05  Konrad Rieck <rieck@cs.tu-berlin.de>

    - added documenation for hash_seed parameters

2009-12-05  Konrad Rieck <rieck@fluffy.my.domain>

    - support changing the seed of MD5 in feature extraction

    - fixes for OpenBSD (including annoying FLT_NONE hack)

    - fixed minimalistic  matlab interface

2009-11-07  Konrad Rieck <rieck@cs.tu-berlin.de>

    - extension and spell fixes to manual

2009-11-07  Konrad Rieck <rieck@node04.ml.tu-berlin.de>

    - openmp enabled by default

2009-11-06  Konrad Rieck <rieck@cs.tu-berlin.de>

    - improved manpage egenration

2009-10-30  Konrad Rieck <rieck@cs.tu-berlin.de>

    - fixes for distribution

    - support to write version in latex

    - added nice output for pod file

    - added pod infavor of man.in

    - fixes to documentation

    - changed creation of man page

    - first paragraph

    - fixed copyright notice

    - minor fixes

2009-10-16  Konrad Rieck <rieck@cs.tu-berlin.de>

    - several improvements of doxygen documentation

2009-10-15  Konrad Rieck <rieck@cs.tu-berlin.de>

    - improved layout wrt assignments vs. classification using prototyes

    - changed file names

    - improved run-time of prototype extraction

    - minor speed improvements

2009-10-14  Konrad Rieck <rieck@cs.tu-berlin.de>

    - linux and matlab fixes

    - improved incremental output

    - first version of incremental analysis

2009-10-13  Konrad Rieck <rieck@cs.tu-berlin.de>

    - long overdue code formating

    - fix to automake

    - minor fixes

    - added classification test case

    - added clustering test case

    - added test case for prototype extraction

    - minor fixes in code layout

2009-10-12  Konrad Rieck <rieck@ubuntu.(none)>

    - fixed configuration script

2009-10-12  Konrad Rieck <rieck@cs.tu-berlin.de>

    - forgot class files. ooops.

    - more doxygen fixes

    - fixed comments

    - improved classification routines

    - sync

    - working on cleaner design of clustering/classificaion/prototype stuff

    - changed load and save function for prototypes

2009-10-10  Konrad Rieck <rieck@cs.tu-berlin.de>

    - first version with clustering, classification and prototypes

    - added support for rejection class in classification

    - added support for standard prototype and reject files

2009-10-09  Konrad Rieck <rieck@ubuntu.(none)>

    - fixed bug in matrix access (thanks to valgrind)

2009-10-09  Konrad Rieck <rieck@cs.tu-berlin.de>

    - more fixes to prototypes and export function

    - radical clean up of unused functions

    - simplified matlab cruft

    - further improvements of clustering and prototype computation

    - slimmed down prototype extraction

    - sanitized text output

    - removed annoying HTML and sandbox dependencies

    - fixes for OSX 10.6

2009-10-08  Konrad Rieck <rieck@cs.tu-berlin.de>

    - first version of clustering (Murtagh's algorithm)

    - minor fixes to configuration

2009-09-23  Konrad Rieck <rieck@node15.ml.tu-berlin.de>


src/md5.c
/*
 * This code implements the MD5 message-digest algorithm.
 * The algorithm is due to Ron Rivest.  This code was
 * written by Colin Plumb in 1993, no copyright is claimed.
 * This code is in the public domain; do with it what you wish.
 *
 * Equivalent code is available from RSA Data Security, Inc.
 * This code has been tested against that, and is equivalent,
 * except that you don't need to include two pages of legalese
 * with every copy.
 *
 * To compute the message digest of a chunk of bytes, declare an
 * MD5Context structure, pass it to MD5Init, call MD5Update as
 * needed on buffers full of bytes, and then call MD5Final, which
 * will fill a supplied 16-byte array with the digest.
 */

#include "config.h"
#include "common.h"
#include "md5.h"

/* External variables */
extern config_t cfg;

/* Global variables for seed */
static uint32_t seed1 = MD5_SEED_NONE;
static uint32_t seed2 = MD5_SEED_NONE;

/*
 * Wrapper for MD5 function compatible to OpenSSL interface.
 * 'buf' needs to provide an allocated array of 16 bytes for 
 * the digest.
 */
void MD5(unsigned char *x, unsigned l, unsigned char *buf)
{
    struct MD5Context ctx;

    MD5Init(&ctx);
    MD5Update(&ctx, x, l);
    MD5Final(buf, &ctx);
}

#ifndef HIGHFIRST
#define byteReverse(buf, len)   /* Nothing */
#else
void byteReverse(unsigned char *buf, unsigned uint32_t longs);

#ifndef ASM_MD5
/*
 * Note: this code is harmless on little-endian machines.
 */
void byteReverse(unsigned char *buf, unsigned uint32_t longs)
{
    uint32_t t;
    do {
        t = (uint32_t) ((unsigned) buf[3] << 8 | buf[2]) << 16 |
            ((unsigned) buf[1] << 8 | buf[0]);
        *(uint32_t *) buf = t;
        buf += 4;
    } while (--longs);
}
#endif
#endif

/*
 * Start MD5 accumulation.  Set bit count to 0 and buffer to mysterious
 * initialization constants.
 */
void MD5Init(struct MD5Context *ctx)
{
    /* Initialize with selected seed values */
    if (seed1 == MD5_SEED_NONE) {
        long long l;
        config_lookup_int(&cfg, "features.hash_seed1", (int *) &l); 
        seed1 = (uint32_t) l;
        config_lookup_int(&cfg, "features.hash_seed2", (int *) &l);    
        seed2 = (uint32_t) l;
    }
    
    ctx->buf[0] = seed1;
    ctx->buf[1] = seed2;
    ctx->buf[2] = 0x98badcfe;
    ctx->buf[3] = 0x10325476;

    ctx->bits[0] = 0;
    ctx->bits[1] = 0;
}

/*
 * Update context to reflect the concatenation of another buffer full
 * of bytes.
 */
void MD5Update(struct MD5Context *ctx, unsigned char const *buf,
               unsigned len)
{
    uint32_t t;

    /* Update bitcount */

    t = ctx->bits[0];
    if ((ctx->bits[0] = t + ((uint32_t) len << 3)) < t)
        ctx->bits[1]++;         /* Carry from low to high */
    ctx->bits[1] += len >> 29;

    t = (t >> 3) & 0x3f;        /* Bytes already in shsInfo->data */

    /* Handle any leading odd-sized chunks */

    if (t) {
        unsigned char *p = (unsigned char *) ctx->in + t;

        t = 64 - t;
        if (len < t) {
            memcpy(p, buf, len);
            return;
        }
        memcpy(p, buf, t);
        byteReverse(ctx->in, 16);
        MD5Transform(ctx->buf, (uint32_t *) ctx->in);
        buf += t;
        len -= t;
    }
    /* Process data in 64-byte chunks */

    while (len >= 64) {
        memcpy(ctx->in, buf, 64);
        byteReverse(ctx->in, 16);
        MD5Transform(ctx->buf, (uint32_t *) ctx->in);
        buf += 64;
        len -= 64;
    }

    /* Handle any remaining bytes of data. */

    memcpy(ctx->in, buf, len);
}

/*
 * Final wrapup - pad to 64-byte boundary with the bit pattern 
 * 1 0* (64-bit count of bits processed, MSB-first)
 */
void MD5Final(unsigned char digest[16], struct MD5Context *ctx)
{
    unsigned count;
    unsigned char *p;

    /* Compute number of bytes mod 64 */
    count = (ctx->bits[0] >> 3) & 0x3F;

    /* Set the first char of padding to 0x80.  This is safe since there is
       always at least one byte free */
    p = ctx->in + count;
    *p++ = 0x80;

    /* Bytes of padding needed to make 64 bytes */
    count = 64 - 1 - count;

    /* Pad out to 56 mod 64 */
    if (count < 8) {
        /* Two lots of padding:  Pad the first block to 64 bytes */
        memset(p, 0, count);
        byteReverse(ctx->in, 16);
        MD5Transform(ctx->buf, (uint32_t *) ctx->in);

        /* Now fill the next block with 56 bytes */
        memset(ctx->in, 0, 56);
    } else {
        /* Pad block to 56 bytes */
        memset(p, 0, count - 8);
    }
    byteReverse(ctx->in, 14);

    /* Append length in bits and transform */
    memcpy (ctx->in + 56, &ctx->bits[0], sizeof(uint32_t));
    memcpy (ctx->in + 60, &ctx->bits[1], sizeof(uint32_t));

    MD5Transform(ctx->buf, (uint32_t *) ctx->in);
    byteReverse((unsigned char *) ctx->buf, 4);
    memcpy(digest, ctx->buf, 16);
    memset(ctx, 0, sizeof(ctx));        /* In case it's sensitive */
}

#ifndef ASM_MD5

/* The four core functions - F1 is optimized somewhat */

/* #define F1(x, y, z) (x & y | ~x & z) */
#define F1(x, y, z) (z ^ (x & (y ^ z)))
#define F2(x, y, z) F1(z, x, y)
#define F3(x, y, z) (x ^ y ^ z)
#define F4(x, y, z) (y ^ (x | ~z))

/* This is the central step in the MD5 algorithm. */
#ifdef __PUREC__
#define MD5STEP(f, w, x, y, z, data, s) \
	( w += f /*(x, y, z)*/ + data,  w = w<<s | w>>(32-s),  w += x )
#else
#define MD5STEP(f, w, x, y, z, data, s) \
	( w += f(x, y, z) + data,  w = w<<s | w>>(32-s),  w += x )
#endif

/*
 * The core of the MD5 algorithm, this alters an existing MD5 hash to
 * reflect the addition of 16 long words of new data.  MD5Update blocks
 * the data and converts bytes into longwords for this routine.
 */
void MD5Transform(uint32_t buf[4], uint32_t const in[16])
{
    register uint32_t a, b, c, d;

    a = buf[0];
    b = buf[1];
    c = buf[2];
    d = buf[3];

#ifdef __PUREC__                /* PureC Weirdness... (GG) */
    MD5STEP(F1(b, c, d), a, b, c, d, in[0] + 0xd76aa478L, 7);
    MD5STEP(F1(a, b, c), d, a, b, c, in[1] + 0xe8c7b756L, 12);
    MD5STEP(F1(d, a, b), c, d, a, b, in[2] + 0x242070dbL, 17);
    MD5STEP(F1(c, d, a), b, c, d, a, in[3] + 0xc1bdceeeL, 22);
    MD5STEP(F1(b, c, d), a, b, c, d, in[4] + 0xf57c0fafL, 7);
    MD5STEP(F1(a, b, c), d, a, b, c, in[5] + 0x4787c62aL, 12);
    MD5STEP(F1(d, a, b), c, d, a, b, in[6] + 0xa8304613L, 17);
    MD5STEP(F1(c, d, a), b, c, d, a, in[7] + 0xfd469501L, 22);
    MD5STEP(F1(b, c, d), a, b, c, d, in[8] + 0x698098d8L, 7);
    MD5STEP(F1(a, b, c), d, a, b, c, in[9] + 0x8b44f7afL, 12);
    MD5STEP(F1(d, a, b), c, d, a, b, in[10] + 0xffff5bb1L, 17);
    MD5STEP(F1(c, d, a), b, c, d, a, in[11] + 0x895cd7beL, 22);
    MD5STEP(F1(b, c, d), a, b, c, d, in[12] + 0x6b901122L, 7);
    MD5STEP(F1(a, b, c), d, a, b, c, in[13] + 0xfd987193L, 12);
    MD5STEP(F1(d, a, b), c, d, a, b, in[14] + 0xa679438eL, 17);
    MD5STEP(F1(c, d, a), b, c, d, a, in[15] + 0x49b40821L, 22);

    MD5STEP(F2(b, c, d), a, b, c, d, in[1] + 0xf61e2562L, 5);
    MD5STEP(F2(a, b, c), d, a, b, c, in[6] + 0xc040b340L, 9);
    MD5STEP(F2(d, a, b), c, d, a, b, in[11] + 0x265e5a51L, 14);
    MD5STEP(F2(c, d, a), b, c, d, a, in[0] + 0xe9b6c7aaL, 20);
    MD5STEP(F2(b, c, d), a, b, c, d, in[5] + 0xd62f105dL, 5);
    MD5STEP(F2(a, b, c), d, a, b, c, in[10] + 0x02441453L, 9);
    MD5STEP(F2(d, a, b), c, d, a, b, in[15] + 0xd8a1e681L, 14);
    MD5STEP(F2(c, d, a), b, c, d, a, in[4] + 0xe7d3fbc8L, 20);
    MD5STEP(F2(b, c, d), a, b, c, d, in[9] + 0x21e1cde6L, 5);
    MD5STEP(F2(a, b, c), d, a, b, c, in[14] + 0xc33707d6L, 9);
    MD5STEP(F2(d, a, b), c, d, a, b, in[3] + 0xf4d50d87L, 14);
    MD5STEP(F2(c, d, a), b, c, d, a, in[8] + 0x455a14edL, 20);
    MD5STEP(F2(b, c, d), a, b, c, d, in[13] + 0xa9e3e905L, 5);
    MD5STEP(F2(a, b, c), d, a, b, c, in[2] + 0xfcefa3f8L, 9);
    MD5STEP(F2(d, a, b), c, d, a, b, in[7] + 0x676f02d9L, 14);
    MD5STEP(F2(c, d, a), b, c, d, a, in[12] + 0x8d2a4c8aL, 20);

    MD5STEP(F3(b, c, d), a, b, c, d, in[5] + 0xfffa3942L, 4);
    MD5STEP(F3(a, b, c), d, a, b, c, in[8] + 0x8771f681L, 11);
    MD5STEP(F3(d, a, b), c, d, a, b, in[11] + 0x6d9d6122L, 16);
    MD5STEP(F3(c, d, a), b, c, d, a, in[14] + 0xfde5380cL, 23);
    MD5STEP(F3(b, c, d), a, b, c, d, in[1] + 0xa4beea44L, 4);
    MD5STEP(F3(a, b, c), d, a, b, c, in[4] + 0x4bdecfa9L, 11);
    MD5STEP(F3(d, a, b), c, d, a, b, in[7] + 0xf6bb4b60L, 16);
    MD5STEP(F3(c, d, a), b, c, d, a, in[10] + 0xbebfbc70L, 23);
    MD5STEP(F3(b, c, d), a, b, c, d, in[13] + 0x289b7ec6L, 4);
    MD5STEP(F3(a, b, c), d, a, b, c, in[0] + 0xeaa127faL, 11);
    MD5STEP(F3(d, a, b), c, d, a, b, in[3] + 0xd4ef3085L, 16);
    MD5STEP(F3(c, d, a), b, c, d, a, in[6] + 0x04881d05L, 23);
    MD5STEP(F3(b, c, d), a, b, c, d, in[9] + 0xd9d4d039L, 4);
    MD5STEP(F3(a, b, c), d, a, b, c, in[12] + 0xe6db99e5L, 11);
    MD5STEP(F3(d, a, b), c, d, a, b, in[15] + 0x1fa27cf8L, 16);
    MD5STEP(F3(c, d, a), b, c, d, a, in[2] + 0xc4ac5665L, 23);

    MD5STEP(F4(b, c, d), a, b, c, d, in[0] + 0xf4292244L, 6);
    MD5STEP(F4(a, b, c), d, a, b, c, in[7] + 0x432aff97L, 10);
    MD5STEP(F4(d, a, b), c, d, a, b, in[14] + 0xab9423a7L, 15);
    MD5STEP(F4(c, d, a), b, c, d, a, in[5] + 0xfc93a039L, 21);
    MD5STEP(F4(b, c, d), a, b, c, d, in[12] + 0x655b59c3L, 6);
    MD5STEP(F4(a, b, c), d, a, b, c, in[3] + 0x8f0ccc92L, 10);
    MD5STEP(F4(d, a, b), c, d, a, b, in[10] + 0xffeff47dL, 15);
    MD5STEP(F4(c, d, a), b, c, d, a, in[1] + 0x85845dd1L, 21);
    MD5STEP(F4(b, c, d), a, b, c, d, in[8] + 0x6fa87e4fL, 6);
    MD5STEP(F4(a, b, c), d, a, b, c, in[15] + 0xfe2ce6e0L, 10);
    MD5STEP(F4(d, a, b), c, d, a, b, in[6] + 0xa3014314L, 15);
    MD5STEP(F4(c, d, a), b, c, d, a, in[13] + 0x4e0811a1L, 21);
    MD5STEP(F4(b, c, d), a, b, c, d, in[4] + 0xf7537e82L, 6);
    MD5STEP(F4(a, b, c), d, a, b, c, in[11] + 0xbd3af235L, 10);
    MD5STEP(F4(d, a, b), c, d, a, b, in[2] + 0x2ad7d2bbL, 15);
    MD5STEP(F4(c, d, a), b, c, d, a, in[9] + 0xeb86d391L, 21);
#else
    MD5STEP(F1, a, b, c, d, in[0] + 0xd76aa478, 7);
    MD5STEP(F1, d, a, b, c, in[1] + 0xe8c7b756, 12);
    MD5STEP(F1, c, d, a, b, in[2] + 0x242070db, 17);
    MD5STEP(F1, b, c, d, a, in[3] + 0xc1bdceee, 22);
    MD5STEP(F1, a, b, c, d, in[4] + 0xf57c0faf, 7);
    MD5STEP(F1, d, a, b, c, in[5] + 0x4787c62a, 12);
    MD5STEP(F1, c, d, a, b, in[6] + 0xa8304613, 17);
    MD5STEP(F1, b, c, d, a, in[7] + 0xfd469501, 22);
    MD5STEP(F1, a, b, c, d, in[8] + 0x698098d8, 7);
    MD5STEP(F1, d, a, b, c, in[9] + 0x8b44f7af, 12);
    MD5STEP(F1, c, d, a, b, in[10] + 0xffff5bb1, 17);
    MD5STEP(F1, b, c, d, a, in[11] + 0x895cd7be, 22);
    MD5STEP(F1, a, b, c, d, in[12] + 0x6b901122, 7);
    MD5STEP(F1, d, a, b, c, in[13] + 0xfd987193, 12);
    MD5STEP(F1, c, d, a, b, in[14] + 0xa679438e, 17);
    MD5STEP(F1, b, c, d, a, in[15] + 0x49b40821, 22);

    MD5STEP(F2, a, b, c, d, in[1] + 0xf61e2562, 5);
    MD5STEP(F2, d, a, b, c, in[6] + 0xc040b340, 9);
    MD5STEP(F2, c, d, a, b, in[11] + 0x265e5a51, 14);
    MD5STEP(F2, b, c, d, a, in[0] + 0xe9b6c7aa, 20);
    MD5STEP(F2, a, b, c, d, in[5] + 0xd62f105d, 5);
    MD5STEP(F2, d, a, b, c, in[10] + 0x02441453, 9);
    MD5STEP(F2, c, d, a, b, in[15] + 0xd8a1e681, 14);
    MD5STEP(F2, b, c, d, a, in[4] + 0xe7d3fbc8, 20);
    MD5STEP(F2, a, b, c, d, in[9] + 0x21e1cde6, 5);
    MD5STEP(F2, d, a, b, c, in[14] + 0xc33707d6, 9);
    MD5STEP(F2, c, d, a, b, in[3] + 0xf4d50d87, 14);
    MD5STEP(F2, b, c, d, a, in[8] + 0x455a14ed, 20);
    MD5STEP(F2, a, b, c, d, in[13] + 0xa9e3e905, 5);
    MD5STEP(F2, d, a, b, c, in[2] + 0xfcefa3f8, 9);
    MD5STEP(F2, c, d, a, b, in[7] + 0x676f02d9, 14);
    MD5STEP(F2, b, c, d, a, in[12] + 0x8d2a4c8a, 20);

    MD5STEP(F3, a, b, c, d, in[5] + 0xfffa3942, 4);
    MD5STEP(F3, d, a, b, c, in[8] + 0x8771f681, 11);
    MD5STEP(F3, c, d, a, b, in[11] + 0x6d9d6122, 16);
    MD5STEP(F3, b, c, d, a, in[14] + 0xfde5380c, 23);
    MD5STEP(F3, a, b, c, d, in[1] + 0xa4beea44, 4);
    MD5STEP(F3, d, a, b, c, in[4] + 0x4bdecfa9, 11);
    MD5STEP(F3, c, d, a, b, in[7] + 0xf6bb4b60, 16);
    MD5STEP(F3, b, c, d, a, in[10] + 0xbebfbc70, 23);
    MD5STEP(F3, a, b, c, d, in[13] + 0x289b7ec6, 4);
    MD5STEP(F3, d, a, b, c, in[0] + 0xeaa127fa, 11);
    MD5STEP(F3, c, d, a, b, in[3] + 0xd4ef3085, 16);
    MD5STEP(F3, b, c, d, a, in[6] + 0x04881d05, 23);
    MD5STEP(F3, a, b, c, d, in[9] + 0xd9d4d039, 4);
    MD5STEP(F3, d, a, b, c, in[12] + 0xe6db99e5, 11);
    MD5STEP(F3, c, d, a, b, in[15] + 0x1fa27cf8, 16);
    MD5STEP(F3, b, c, d, a, in[2] + 0xc4ac5665, 23);

    MD5STEP(F4, a, b, c, d, in[0] + 0xf4292244, 6);
    MD5STEP(F4, d, a, b, c, in[7] + 0x432aff97, 10);
    MD5STEP(F4, c, d, a, b, in[14] + 0xab9423a7, 15);
    MD5STEP(F4, b, c, d, a, in[5] + 0xfc93a039, 21);
    MD5STEP(F4, a, b, c, d, in[12] + 0x655b59c3, 6);
    MD5STEP(F4, d, a, b, c, in[3] + 0x8f0ccc92, 10);
    MD5STEP(F4, c, d, a, b, in[10] + 0xffeff47d, 15);
    MD5STEP(F4, b, c, d, a, in[1] + 0x85845dd1, 21);
    MD5STEP(F4, a, b, c, d, in[8] + 0x6fa87e4f, 6);
    MD5STEP(F4, d, a, b, c, in[15] + 0xfe2ce6e0, 10);
    MD5STEP(F4, c, d, a, b, in[6] + 0xa3014314, 15);
    MD5STEP(F4, b, c, d, a, in[13] + 0x4e0811a1, 21);
    MD5STEP(F4, a, b, c, d, in[4] + 0xf7537e82, 6);
    MD5STEP(F4, d, a, b, c, in[11] + 0xbd3af235, 10);
    MD5STEP(F4, c, d, a, b, in[2] + 0x2ad7d2bb, 15);
    MD5STEP(F4, b, c, d, a, in[9] + 0xeb86d391, 21);
#endif

    buf[0] += a;
    buf[1] += b;
    buf[2] += c;
    buf[3] += d;
}

#endif
